decisions/0009-reorganize-code-for-interacting-with-kubernetes.md

```diff
@@ -0,0 +1,98 @@
+     request
+   - Any other data that has been added to the existing resource (and cannot be
+     derived from the CC model) will be removed.
+
+Having a separate translator object could help avoid code drift should we have 2
+or more Ruby classes that are building/manipulating the same type of Kubernetes
+resource. However, in the current codebase there is exactly one Ruby class that
+manages each type of k8s resource. If this changes we can reevaluate this
+decision.
+
+### Propagate the Kubernetes resource JSON
+ADR 6 proposed having an interface like `kubectl apply`: a single k8s API client
+that takes a k8s resource config and sends it to the correct API endpoint. None
+of the 3 ruby clients [mentioned in k8s docs] support this workflow. They all
+either provide a separate set of methods for each resource (e.g. `create_image`)
+or require you to scope your get/update/etc to a specific resource (e.g.
+`client.api('v1').resource('services').create_resource(service)`).
+
+We could write our own code to make this flow possible, but the benefit of that
+isn't clear. If anything, having different methods to reflect each
+(verb, resource-type) tuple makes it easier to test our interaction with the k8s
+API in CC (via `expect(client).to(have_received(:create_image))`). It should
+also yield a clear "No method" error if we attempt to use a verb or
+resource-type that isn't supported.
+
+
+## Decision
+
+1. Create a single Ruby class to manage interactions with the k8s API. This
+   class will handle translating Kubeclient errors into CC errors and little
+   else. We will continue the existing pattern of a separate method for each
+   verb + resource-type combination (e.g. `#update_route`)
+   - This class is currently called `Kubernetes::ApiClient`
+
+2. Rather than creating a separate translation class, we will maintain the
+   current pattern:
+   - Each type of k8s resource is managed by exactly one Ruby class
+   - That class contains the logic for creating or modifying the k8s resource
+     config based on CC models
+   - That class delegates to `Kubernetes::ApiClient` to interact with the k8s
+     API and handle error translation
+   - Current examples of classes that do this: 
+     - [`Kpack::Stager`](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/lib/cloud_controller/kpack/stager.rb)
+     - [`VCAP::CloudController::KpackBuildpackListFetcher`](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/app/fetchers/kpack_buildpack_list_fetcher.rb)
+     - [`Kubernetes::RouteResourceManager`](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/app/fetchers/kpack_buildpack_list_fetcher.rb) (formerly `RouteCrdClient`)
+
+3. Move all k8s-specific classes under a single namespace (e.g. `Kubernetes` or
+   `VCAP::CloudController::K8s`)
+4. Move all classes that are specific to bosh/VM-based deployments of CF into
+   their own namespace as well (e.g. `Vm`, `Bosh`, `Classic`, or `Legacy`)
```

Comment:
Point 4 here is pretty broad. I'm in favor of it aspirationally, but I don't think it can be done all in one change. 

Some of these "legacy" classes live under `VCAP::CloudController::Diego`, but a lot of the code in that module is actually still in use in cf-for-k8s despite the absence of Diego. I think all of the top level `::Diego` code is BOSH only. Attempting to untangle all that when we don't have context or eyes on it currently might be risky.

Reply:
Thank you for the context. I definitely think of this as a nice-to-have and less important than the other items here. That said, untangling what "Diego" code is actually generic code would make the codebase clearer, so may be worthwhile.